R/xml.R

```diff
@@ -0,0 +1,576 @@
+# xml.R
+
+#' create_xml
+#'
+#' The basis to define how to convert data to an XML file.  This method
+#' simple requires the name to save the XML file as and optionally the
+#' model interface "header" file that defines the transformation lookup
+#' to go from tabular data to hierarchical.  The result of this should be
+#' used in a dplyr pipeline with one or more calls to \code{\link{add_xml_data}}
+#' to add the data to convert and finally ending with \code{\link{run_xml_conversion}}
+#' to run the conversion.
+#'
+#' @param xml_file The name to save the XML file to
+#' @param mi_header The model interface "header".  This will default to the one
+#' included in this package
+#' @return A "data structure" to hold the various parts needed to run the model
+#' interface CSV to XML conversion.
+#' @export
+create_xml <- function(xml_file, mi_header = NULL) {
+  if(is.null(mi_header)) {
+    mi_header <- system.file("extdata/mi_headers", "ModelInterface_headers.txt",
+                             package = "gcamdata")
+  }
+
+  xml_obj <- list(xml_file = xml_file,
+                  mi_header = mi_header,
+                  data_tables = list())
+  xml_obj <- add_flags(xml_obj, FLAG_XML)
+
+  xml_obj
+}
+
+#' add_xml_data
+#'
+#' Add a table to include for conversion to XML.  We need the tibble to convert
+#' and a header tag which can be looked up in the header file to convert the
+#' tibble.  This method is meant to be included in a pipeline between calls of
+#' \code{\link{create_xml}} and \code{\link{run_xml_conversion}}.
+#'
+#' @param dot The current state of the pipeline started from \code{create_xml}
+#' @param data The tibble of data to add to the conversion
+#' @param header The header tag to can be looked up in the header file to
+#' convert \code{data}
+#' @param column_order_lookup A tag that can be used to look up \code{LEVEL2_DATA_NAMES}
+#' to reorder the columns of data before XML conversion to ensure the correspond
+#' with the ModelInterface header.  Note by default the \code{header} is used and if
+#' given \code{NULL} no column reordering will be done.
+#' @return A "data structure" to hold the various parts needed to run the model
+#' interface CSV to XML conversion.
+#' @author PP March 2017
+#' @export
+add_xml_data <- function(dot, data, header, column_order_lookup = header) {
+  # Users can skip column reordering by passing NULL as the argument value.
+  if(!is.null(column_order_lookup)) {
+    data <- data[, LEVEL2_DATA_NAMES[[column_order_lookup]] ]
+  }
+
+  curr_table <- list(data = data, header = header)
+  dot$data_tables[[length(dot$data_tables)+1]] <- curr_table
+
+  dot
+}
+
+# Note: we have put the definition of run_xml_conversion inside of the "closure"
+# make_run_xml_conversion so that we may stash the XML_WARNING_GIVEN flag in an
+# environment that can only be reached by run_xml_conversion.
```

Comment:
Why do we need to do this?  (It's obvious if you read the code, but it's a little mysterious if you're just reading the comment.)